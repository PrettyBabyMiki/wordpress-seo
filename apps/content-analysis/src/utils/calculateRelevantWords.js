import { get, isEqual, take } from "lodash-es";
import getLanguage from "yoastsrc/helpers/getLanguage";
import getWords from "yoastsrc/stringProcessing/getWords";
import {
	collapseRelevantWordsOnStem,
	getRelevantWords,
	getRelevantWordsFromPaperAttributes,
	getRelevantCombinations,
	sortCombinations,
} from "yoastsrc/stringProcessing/relevantWords";
import { getSubheadingsTopLevel, removeSubheadingsTopLevel } from "yoastsrc/stringProcessing/getSubheadings";
import { retrieveAbbreviations } from "../../../../packages/yoastseo/src/stringProcessing/relevantWords";
import getMorphologyData from "./getMorphologyData";

const morphologyData = getMorphologyData();

// Cache the relevant words.
let previousRelevantWordsInternalLinking = {
	text: "",
	locale: "en_US",
	description: "",
	keyword: "",
	synonyms: "",
	title: "",
	data: {},
};

let previousRelevantWordsInsights = {
	text: "",
	locale: "en_US",
	data: {},
};

/**
 * Rounds number to four decimals.
 *
 * @param {number} number The number to be rounded.
 *
 * @returns {number} The rounded number.
 */
function formatNumber( number ) {
	if ( Math.round( number ) === number ) {
		return number;
	}

	return Math.round( number * 10000 ) / 10000;
}

/**
 * Calculates all properties for the relevant word objects.
 *
 * @param {Paper}   paper           The paper to analyse.
 * @param {boolean} useAttributes   Whether text from attributes should also be considered for relevant words.
 *
 * @returns {Object} The relevant word objects.
 */
function calculateRelevantWords( paper, useAttributes ) {
	let text = paper.text;
	const language = getLanguage( paper.locale );
	const languageMorphologyData = get( morphologyData, language, false );

	const subheadings = getSubheadingsTopLevel( text ).map( subheading => subheading[ 2 ] );

	const attributes = useAttributes ? [
		paper.keyword,
		paper.synonyms,
		paper.description,
		paper.title,
		subheadings.join( " " ),
	] : [];

	text = useAttributes ? removeSubheadingsTopLevel( text ) : text;

	const abbreviations = retrieveAbbreviations( text.concat( attributes.join( " " ) ) );

	const relevantWordsFromText = getRelevantWords( text, abbreviations, language, languageMorphologyData );

	let relevantWordsFromPaperAttributes = [];

	if ( useAttributes ) {
		relevantWordsFromPaperAttributes = getRelevantWordsFromPaperAttributes(
			attributes,
			abbreviations,
			language,
			languageMorphologyData
		);

		/*
		 * Analogous to the research src/researches/relevantWords.js, all relevant words that come from paper attributes
		 * (and not from text) get a times-3 number of occurrences to support the idea that they are more important than
		 * the words coming from the text. For instance, if a word occurs twice in paper attributes it receives
		 * number_of_occurrences = 6.
		 */
		relevantWordsFromPaperAttributes.forEach( relevantWord => relevantWord.setOccurrences( relevantWord.getOccurrences() * 3 ) );
	}

	const collapsedWords = collapseRelevantWordsOnStem( relevantWordsFromPaperAttributes.concat( relevantWordsFromText ) );
	sortCombinations( collapsedWords );

	/*
	 * Analogous to the research src/researches/relevantWords.js, we limit the number of relevant words in consideration
	 * to 100, i.e. we take 100 first relevant words from the list sorded by number of occurrences first and then
	 * alphabetically.
	 */
	const relevantWords = take( getRelevantCombinations( collapsedWords ), 100 );

	return relevantWords.map( ( word ) => {
		return {
			word: word.getWord(),
			stem: word.getStem(),
			occurrences: word.getOccurrences(),
		};
	} );
}

/**
 * Retrieves the relevant words for Internal linking. Uses cached version when possible.
 *
 * @param {Paper} paper   The paper to get relevant words for.
 *
 * @returns {Object} The relevant words.
 */
export function relevantWordsForInternalLinking( paper ) {
	const text = paper.text;
	const locale = paper.locale;
	const description = paper.description;
	const keyword = paper.keyword;
	const synonyms = paper.synonyms;
	const title = paper.title;

	if (
		! isEqual( text, previousRelevantWordsInternalLinking.text ) ||
		! isEqual( locale, previousRelevantWordsInternalLinking.locale ) ||
		! isEqual( description, previousRelevantWordsInternalLinking.description ) ||
		! isEqual( keyword, previousRelevantWordsInternalLinking.keyword ) ||
		! isEqual( synonyms, previousRelevantWordsInternalLinking.synonyms ) ||
		! isEqual( title, previousRelevantWordsInternalLinking.title )
	) {
		previousRelevantWordsInternalLinking = {
			text,
			locale,
			description,
			keyword,
			synonyms,
			title,
			data: calculateRelevantWords( paper, true ),
		};
	}
	return previousRelevantWordsInternalLinking.data;
}

/**
 * Retrieves the relevant words for Insights (takes into consideration only the text itself, not attributes).
 * Uses cached version when possible.
 *
 * @param {Paper} paper   The paper to get relevant words for.
 *
 * @returns {Object} The relevant words.
 */
export function relevantWordsForInsights( paper ) {
	const text = paper.text;
	const locale = paper.locale;

	if (
		! isEqual( text, previousRelevantWordsInsights.text ) ||
		! isEqual( locale, previousRelevantWordsInsights.locale )
	) {
		previousRelevantWordsInsights = {
			text,
			locale,
			data: calculateRelevantWords( paper, false ),
		};
	}
	return previousRelevantWordsInsights.data;
}

