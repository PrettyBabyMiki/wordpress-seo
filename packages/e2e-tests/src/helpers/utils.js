/**
 * WordPress e2e utilities
 */
import {
    installPlugin,
    activatePlugin,
    visitAdminPage,
    pressKeyWithModifier,
} from "@wordpress/e2e-test-utils";

import { addQueryArgs } from '@wordpress/url';

export const createCustomPostType = async( slug ) => {
   
}