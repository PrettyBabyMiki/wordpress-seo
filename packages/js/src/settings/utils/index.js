import { filter, isArray, isObject, join, reduce } from "lodash";

/**
 * Determines if the given value should flatten.
 * @param {*} value The value.
 * @returns {boolean} True if the value is an object or an array.
 */
const shouldFlatten = value => isObject( value ) || isArray( value );

/**
 * Flattens nested object to a single level. Props will be keyed by dot syntax path.
 * @param {Object} object The object to flatten.
 * @param {string} parentPath The parent object path.
 * @param {function} getShouldFlatten Determines whether the entry should be flattened.
 * @returns {Object} Flattened object.
 */
export const flattenObject = ( object, parentPath = "", getShouldFlatten = shouldFlatten ) => reduce(
	object,
	( acc, value, key ) => {
		const flatKey = join( filter( [ parentPath, key ], Boolean ), "." );

		return getShouldFlatten( value, key ) ? {
			...acc,
			...flattenObject( value, flatKey, getShouldFlatten ),
		} : {
			...acc,
			[ flatKey ]: value,
		};
	},
	{}
);
