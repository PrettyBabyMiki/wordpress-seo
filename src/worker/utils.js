/**
 * Encodes the payload for transfer.
 *
 * @param {Object} payload The payload.
 *
 * @returns {string} Transferable string.
 */
export function encodePayload( payload ) {
	return JSON.stringify( payload );
}

/**
 * Decodes the payload for use.
 *
 * @param {string} payload The encoded payload.
 *
 * @returns {any} The parsed result.
 */
export function decodePayload( payload ) {
	try {
		return JSON.parse( payload );
	} catch( error ) {
		return payload;
	}
}
