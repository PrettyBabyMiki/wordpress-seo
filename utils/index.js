import { makeOutboundLink } from "./makeOutboundLink";
import { getFeed } from "./getFeed";

export default {
	getFeed,
	makeOutboundLink,
};
