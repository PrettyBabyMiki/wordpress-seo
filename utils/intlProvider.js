import React from "react";
import renderer from "react-test-renderer";
import { IntlProvider } from "react-intl";

export const createComponentWithIntl = ( children, props = { locale: "en" }, renderOptions = {} ) => {
	return renderer.create(
		<IntlProvider { ...props }>
			{ children }
		</IntlProvider>,
		renderOptions
	);
};
