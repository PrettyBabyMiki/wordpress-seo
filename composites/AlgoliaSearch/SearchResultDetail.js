import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { defineMessages, injectIntl, intlShape } from "react-intl";

import ArticleContent from "./ArticleContent";
import { YoastButton } from "../Plugin/Shared/components/YoastButton";
import { YoastLinkButton } from "../Plugin/Shared/components/YoastLinkButton";
import SvgIcon from "../Plugin/Shared/components/SvgIcon";
import { makeOutboundLink } from "../../utils/makeOutboundLink";
import breakpoints from "../../style-guide/responsive-breakpoints.json";
import colors from "../../style-guide/colors.json";

const messages = defineMessages( {
	searchResult: {
		id: "searchResultDetail.searchResult",
		defaultMessage: "Search result",
	},
	openButton: {
		id: "searchResultDetail.openButton",
		defaultMessage: "View in KB",
	},
	openButtonLabel: {
		id: "searchResultDetail.openButtonLabel",
		defaultMessage: "Open the knowledge base article in a new window or read it in the iframe below",
	},
	backButton: {
		id: "searchResultDetail.backButton",
		defaultMessage: "Go back",
	},
	backButtonLabel: {
		id: "searchResultDetail.backButtonLabel",
		defaultMessage: "Go back to the search results",
	},
	iframeTitle: {
		id: "searchResultDetail.iframeTitle",
		defaultMessage: "Knowledge base article",
	},
} );

const Detail = styled.section`
	outline: none;

	@media screen and ( max-width: ${ breakpoints.mobile } ) {
		margin: 0 -16px;
	}
`;

const Nav = styled.nav`
	padding: 0 16px 16px;
`;

const RightYoastLinkButton = makeOutboundLink( styled( YoastLinkButton )`
	float: right;
` );

/**
 * Create the JSX to render the SearchResultDetail component.
 *
 * @param {object} props The React props.
 *
 * @returns {ReactElement} A SearchResultDetail component.
 */
class SearchResultDetail extends React.Component {
	/**
	 * Creates navigational elements.
	 *
	 * @returns {ReactElement} The rendered element.
	 */
	createNavigation() {
		const formatMessage = this.props.intl.formatMessage;
		const openButtonText = formatMessage( messages.openButton );
		const openButtonLabel = formatMessage( messages.openButtonLabel );
		const backButtonText = formatMessage( messages.backButton );
		const backButtonLabel = formatMessage( messages.backButtonLabel );
		return (
			<Nav>
				<YoastButton aria-label={ backButtonLabel } onClick={ this.props.onBackButtonClicked }>
					<SvgIcon
						size="24px"
						color={ colors.$color_white }
						icon="angle-left"
					/>
					{ backButtonText }
				</YoastButton>
				<RightYoastLinkButton
					href={ this.props.post.permalink }
					aria-label={ openButtonLabel }
					target="_blank"
					rel="noopener noreferrer"
				>
					{ openButtonText }
					<SvgIcon
						size="24px"
						color={ colors.$color_white }
						icon="angle-right"
					/>
				</RightYoastLinkButton>
			</Nav>
		);
	}

	/**
	 * Renders the search result detail.
	 *
	 * @returns {ReactElement} The rendered element.
	 */
	render() {
		const formatMessage = this.props.intl.formatMessage;
		const searchResulLabel = formatMessage( messages.searchResult );
		const iframeTitle = formatMessage( messages.iframeTitle );
		return (
			<Detail
				aria-label={ searchResulLabel }
				tabIndex="-1"
				innerRef={ ( el ) => {
					this.detailWrapper = el;
				} }
			>
				{ this.createNavigation() }
				<ArticleContent post={ this.props.post } title={ iframeTitle } />
			</Detail>
		);
	}

	/**
	 * When the component mounts, set focus on the search detail wrapper.
	 *
	 * @returns {void}
	 */
	componentDidMount() {
		if ( this.detailWrapper !== null ) {
			this.detailWrapper.focus();
		}
	}
}

SearchResultDetail.propTypes = {
	post: PropTypes.object.isRequired,
	onBackButtonClicked: PropTypes.func.isRequired,
	intl: intlShape.isRequired,
};

export default injectIntl( SearchResultDetail );
