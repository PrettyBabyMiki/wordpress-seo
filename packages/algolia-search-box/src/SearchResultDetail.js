/* External imports */
import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { defineMessages, injectIntl, intlShape } from "react-intl";

/* Yoast dependencies */
import { YoastButton, SvgIcon, YoastLinkButton } from "@yoast/components";
import { makeOutboundLink } from "@yoast/helpers";
import { colors, breakpoints } from "@yoast/style-guide";

/* Internal imports */
import ArticleContent from "./ArticleContent";

const messages = defineMessages( {
	searchResult: {
		id: "searchResultDetail.searchResult",
		defaultMessage: "Search result",
	},
	openButton: {
		id: "searchResultDetail.openButton",
		defaultMessage: "View in KB",
	},
	openButtonLabel: {
		id: "searchResultDetail.openButtonLabel",
		defaultMessage: "Open the knowledge base article in a new window or read it in the iframe below",
	},
	backButton: {
		id: "searchResultDetail.backButton",
		defaultMessage: "Go back",
	},
	backButtonLabel: {
		id: "searchResultDetail.backButtonLabel",
		defaultMessage: "Go back to the search results",
	},
	iframeTitle: {
		id: "searchResultDetail.iframeTitle",
		defaultMessage: "Knowledge base article",
	},
} );

const Detail = styled.section`
	outline: none;

	@media screen and ( max-width: ${ breakpoints.mobile } ) {
		margin: 0 -16px;
	}
`;

const Nav = styled.nav`
	padding: 0 16px 16px;
`;

const RightYoastLinkButton = makeOutboundLink( styled( YoastLinkButton )`
	float: right;
` );

/**
 * Create the JSX to render the SearchResultDetail component.
 *
 * @param {object} props The React props.
 *
 * @returns {ReactElement} A SearchResultDetail component.
 */
class SearchResultDetail extends React.Component {
	/**
	 * Creates navigational elements.
	 *
	 * @returns {ReactElement} The rendered element.
	 */
	createNavigation() {
		const formatMessage = this.props.intl.formatMessage;
		const openButtonText = formatMessage( messages.openButton );
		const openButtonLabel = formatMessage( messages.openButtonLabel );
		const backButtonText = formatMessage( messages.backButton );
		const backButtonLabel = formatMessage( messages.backButtonLabel );
		return (
			<Nav>
				<YoastButton aria-label={ backButtonLabel } onClick={ this.props.onBackButtonClicked }>
					<SvgIcon
						size="24px"
						color={ colors.$color_white }
						icon="angle-left"
					/>
					{ backButtonText }
				</YoastButton>
				<RightYoastLinkButton
					href={ this.props.post.permalink }
					aria-label={ openButtonLabel }
				>
					{ openButtonText }
					<SvgIcon
						size="24px"
						color={ colors.$color_white }
						icon="angle-right"
					/>
				</RightYoastLinkButton>
			</Nav>
		);
	}

	/**
	 * Renders the search result detail.
	 *
	 * @returns {ReactElement} The rendered element.
	 */
	render() {
		const formatMessage = this.props.intl.formatMessage;
		const searchResulLabel = formatMessage( messages.searchResult );
		const iframeTitle = formatMessage( messages.iframeTitle );
		return (
			<Detail
				aria-label={ searchResulLabel }
				tabIndex="-1"
				ref={ ( el ) => {
					this.detailWrapper = el;
				} }
			>
				{ this.createNavigation() }
				<ArticleContent post={ this.props.post } title={ iframeTitle } />
			</Detail>
		);
	}

	/**
	 * When the component mounts, set focus on the search detail wrapper.
	 *
	 * @returns {void}
	 */
	componentDidMount() {
		if ( this.detailWrapper !== null ) {
			this.detailWrapper.focus();
		}
	}
}

SearchResultDetail.propTypes = {
	post: PropTypes.object.isRequired,
	onBackButtonClicked: PropTypes.func.isRequired,
	intl: intlShape.isRequired,
};

export default injectIntl( SearchResultDetail );
