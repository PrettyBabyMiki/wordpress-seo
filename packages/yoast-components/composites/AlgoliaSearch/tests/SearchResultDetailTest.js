import React from "react";

import { createComponentWithIntl } from "../../../utils/intlProvider";
import SearchResultDetail from "../SearchResultDetail.js";
const post = { permalink: "https://kb.yoast.com/kb/passive-voice/", postTitle: "Post Title", objectID: 1 };

test( "the SearchResultDetail component matches the snapshot", () => {
	const component = createComponentWithIntl(
		<SearchResultDetail post={ post } showDetail={ () => {} } onBackButtonClicked={ () => {} } iframeTitle="Title" />
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
