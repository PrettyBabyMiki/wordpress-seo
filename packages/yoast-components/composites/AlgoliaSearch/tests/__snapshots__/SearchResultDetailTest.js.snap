// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the SearchResultDetail component matches the snapshot 1`] = `
.c8 {
  width: 100%;
  height: 600px;
  border: none;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
  min-height: 48px;
  margin: 0;
  padding: 0 16px;
  padding: 0px 16px;
  border: 0;
  border-radius: 4px;
  box-sizing: border-box;
  font: 400 14px/24px "Open Sans",sans-serif;
  text-transform: uppercase;
  box-shadow: 0 2px 8px 0 rgba( 0,0,0,0.3 );
  -webkit-transition: box-shadow 150ms ease-out;
  transition: box-shadow 150ms ease-out;
}

.c2:hover,
.c2:focus,
.c2:active {
  box-shadow: 0 4px 10px 0 rgba( 0,0,0,0.2 ),inset 0 0 0 100px rgba( 0,0,0,0.1 );
  color: #fff;
}

.c2:active {
  -webkit-transform: translateY( 1px );
  -ms-transform: translateY( 1px );
  transform: translateY( 1px );
  box-shadow: none;
}

.c3 {
  color: #fff;
  background: #64a60a;
  min-width: 152px;
  text-shadow: 0 0 2px #000;
  overflow: visible;
  cursor: pointer;
}

.c3::-moz-focus-inner {
  border-width: 0;
}

.c3 span {
  display: inherit;
  -webkit-align-items: inherit;
  -webkit-box-align: inherit;
  -ms-flex-align: inherit;
  align-items: inherit;
  -webkit-box-pack: inherit;
  -webkit-justify-content: inherit;
  -ms-flex-pack: inherit;
  justify-content: inherit;
  width: 100%;
}

.c6 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #fff;
  background: #64a60a;
  min-width: 152px;
  text-shadow: 0 0 2px #000;
}

.c7 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  outline: none;
}

.c1 {
  padding: 0 16px 16px;
}

.c5 {
  float: right;
}

.c4 {
  width: 24px;
  height: 24px;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

@media all and (-ms-high-contrast:none),(-ms-high-contrast:active) {
  .c2::after {
    display: inline-block;
    content: "";
    min-height: 48px;
  }
}

@media screen and ( max-width:768px ) {
  .c0 {
    margin: 0 -16px;
  }
}

<section
  aria-label="Search result"
  className="c0"
  tabIndex="-1"
>
  <nav
    className="c1"
  >
    <button
      aria-expanded={undefined}
      className="c2 YoastButton-hNXTAd c3"
      onClick={[Function]}
      type="button"
    >
      <span>
        <svg
          aria-hidden={true}
          className="yoast-svg-icon yoast-svg-icon-angle-left c4"
          fill="#fff"
          focusable="false"
          role="img"
          size="24px"
          viewBox="0 0 1792 1792"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1203 544q0 13-10 23l-393 393 393 393q10 10 10 23t-10 23l-50 50q-10 10-23 10t-23-10l-466-466q-10-10-10-23t10-23l466-466q10-10 23-10t23 10l50 50q10 10 10 23z"
          />
        </svg>
        Go back
      </span>
    </button>
    <a
      aria-label="Open the knowledge base article in a new window or read it in the iframe below"
      className="c5 c2 c6"
      href="https://kb.yoast.com/kb/passive-voice/"
      rel="noopener noreferrer"
      target="_blank"
    >
      View in KB
      <svg
        aria-hidden={true}
        className="yoast-svg-icon yoast-svg-icon-angle-right c4"
        fill="#fff"
        focusable="false"
        role="img"
        size="24px"
        viewBox="0 0 1792 1792"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1171 960q0 13-10 23l-466 466q-10 10-23 10t-23-10l-50-50q-10-10-10-23t10-23l393-393-393-393q-10-10-10-23t10-23l50-50q10-10 23-10t23 10l466 466q10 10 10 23z"
        />
      </svg>
      <span
        className="c7"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </nav>
  <iframe
    className="kb-search-content-frame c8"
    src="https://kb.yoast.com/kb/passive-voice/amp/?source=wpseo-kb-search"
    title="Knowledge base article"
  />
</section>
`;
